Source/ASDisplayNode.mm

```diff
@@ -948,9 +948,29 @@ - (ASLayout *)calculateLayoutThatFits:(ASSizeRange)constrainedSize
                      restrictedToSize:(ASLayoutElementSize)size
                  relativeToParentSize:(CGSize)parentSize
 {
+  // Use a pthread specific to mark when this method is called re-entrant on same thread.
+  // We only want one calculateLayout signpost interval per thread.
+  // This is fast enough to do it unconditionally.
+  static pthread_key_t calculateLayoutKey;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    pthread_key_create(&calculateLayoutKey, NULL);
+  });
+  BOOL isNested = pthread_getspecific(calculateLayoutKey) != NULL;
+  if (!isNested) {
+    pthread_setspecific(calculateLayoutKey, kCFBooleanTrue);
```

Comment:
Good point, thanks. To split the difference, I created a `ASPthreadStaticKey(dtor)` macro that lets you easily create a static `pthread_key_t` with a given destructor, and used it here. The general CreateOnce macro is great but I wanted something more specialized for this.